src/app/AttributePersistenceProvider.h

```diff
@@ -56,17 +61,171 @@ class AttributePersistenceProvider
-     * @param [in]     aMetadata the attribute metadata, as a convenience.
+     * @param [in]     aType the attribute type.
+     * @param [in]     aSize the attribute size.
      * @param [in,out] aValue where to place the data.  The size of the buffer
-     *                 will be equal to `size` member of aMetadata.
+     *                 will be equal to `size`.
      *
      *                 The data is expected to be in native endianness for
      *                 integers and floats.  For strings, see the string
      *                 representation description in the WriteValue
      *                 documentation.
      */
-    virtual CHIP_ERROR ReadValue(const ConcreteAttributePath & aPath, const EmberAfAttributeMetadata * aMetadata,
+    virtual CHIP_ERROR ReadValue(const ConcreteAttributePath & aPath, EmberAfAttributeType aType, size_t aSize,
                                  MutableByteSpan & aValue) = 0;
+
+    /**
+     * Get the KVS representation of null for the given type.
+     * @tparam T The type for which the null representation should be returned.
+     * @return A value of type T that in the KVS represents null.
+     */
+    template <typename T, std::enable_if_t<std::is_same<bool, T>::value, bool> = true>
+    static uint8_t GetNullValueForNullableType()
+    {
+        return 0xff;
+    }
+
+    /**
+     * Get the KVS representation of null for the given type.
+     * @tparam T The type for which the null representation should be returned.
+     * @return A value of type T that in the KVS represents null.
+     */
+    template <typename T, std::enable_if_t<std::is_unsigned<T>::value && !std::is_same<bool, T>::value, bool> = true>
+    static T GetNullValueForNullableType()
+    {
+        T nullValue = 0;
+        nullValue   = T(~nullValue);
+        return nullValue;
+    }
+
+    /**
+     * Get the KVS representation of null for the given type.
+     * @tparam T The type for which the null representation should be returned.
+     * @return A value of type T that in the KVS represents null.
+     */
+    template <typename T, std::enable_if_t<std::is_signed<T>::value && !std::is_same<bool, T>::value, bool> = true>
+    static T GetNullValueForNullableType()
+    {
+        T shiftBit = 1;
+        return T(shiftBit << ((sizeof(T) * 8) - 1));
+    }
+
+    // The following API provides helper functions to simplify the access of commonly used types.
+    // The API may not be complete.
+    // Currently implemented write and read types are: uint8_t, uint16_t, uint32_t, unit64_t and
+    // their nullable varieties, and bool.
+
+    /**
+     * Write an attribute value of type intX, uintX or bool to non-volatile memory.
+     *
+     * @param [in] aPath the attribute path for the data being written.
+     * @param [in] aValue the data to write.
+     */
+    template <typename T, std::enable_if_t<std::is_integral<T>::value, bool> = true>
```

Comment:
Shouldn't this have tested for T being unsigned too, if we are going to assume below that it is?  (But watch out for bool....)

Reply:
The `std::is_integral` will return `true` if `T` is a signed or unsigned integer. According to [this documentation](https://en.cppreference.com/w/cpp/types/is_integral) at least.

Reply:
I know it will; my point is that the actual code in there ends up assuming the value is unsigned, right?

Reply:
So concretely.  Say this is invoked with T == int8_t, aValue == -1.

We will set up a 1-byte buffer.

We will then proceed to do `w.EndianPut(uint64_t(aValue), 1);` on the buffer writer.  But for a signed value we should be doing `w.EndianPutSigned`, I would think.  So we should have separate things for signed vs unsigned here @hicklin, instead of assuming things that might or might not be true about the implementation details of the buffer writer.